### PR TITLE
Optimize eigen decomposition for Hill tensor

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
- Optimizes the eigen decomposition in the Hill tensor material classes. 
- Utilizes the block diagonal decomposition approach. 

Refs #16016

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C [hunk 1]
```diff
@@ -201,16 +201,40 @@ ADHillElastoPlasticityStressUpdate::computeElasticityTensorEigenDecomposition()
...
-  auto lambda_b = es_b.eigenvalues();
-  auto v_b = es_b.eigenvectors();
-  for (unsigned int index_i = 0; index_i < dimension; index_i++)
-    _b_eigenvalues[_qp](index_i) = lambda_b(index_i);
-
-  for (unsigned int index_i = 0; index_i < dimension; index_i++)
-    for (unsigned int index_j = 0; index_j < dimension; index_j++)
-      _b_eigenvectors[_qp](index_i, index_j) = v_b(index_i, index_j);
+  Eigen::SelfAdjointEigenSolver<AnisotropyMatrixRealBlock> es_b_left(B_eigen.block<3, 3>(0, 0));
+  Eigen::SelfAdjointEigenSolver<AnisotropyMatrixRealBlock> es_b_right(B_eigen.block<3, 3>(3, 3));
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:205]: isn't that block diagonal anyways?

File: modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C [hunk 2]
```diff
@@ -218,16 +218,30 @@ ADHillPlasticityStressUpdate::computeHillTensorEigenDecomposition(const ADDenseM
...
+  _eigenvectors_hill(0, 2) = v(0, 2);
+  _eigenvectors_hill(1, 0) = v(1, 0);
+  _eigenvectors_hill(1, 1) = v(1, 1);
+  _eigenvectors_hill(1, 2) = v(1, 2);
+  _eigenvectors_hill(2, 0) = v(2, 0);
+  _eigenvectors_hill(2, 1) = v(2, 1);
+  _eigenvectors_hill(2, 2) = v(2, 2);
+  _eigenvectors_hill(3, 3) = 1.0;
+  _eigenvectors_hill(4, 4) = 1.0;
+  _eigenvectors_hill(5, 5) = 1.0;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:244]: apparently so.

So, no need to solve anything. The Eigenvalues are the values on the diagonal!

Comment (SudiptaBiswas) [reply modules/tensor_mechanics/src/materials/ADHillPlasticityStressUpdate.C:244]: Yeah, for the bottom part.

File: modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C (hunk 1 above)
Comment (SudiptaBiswas) [reply modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:205]: Not in this case. Here, `B_eigen` is Hill tesnor multiplied with the eigen vectors of the elasticity tensor. For the test cases, the top left part was rather a diagonal matrix. But that may not be always true. So, I did the decomposition for both diagonal blocks.

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:205]: But `B_eigen.block<3, 3>(3, 0)` and `B_eigen.block<3, 3>(0, 3)` are still 0?

Comment (SudiptaBiswas) [reply modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:205]: Yes, they are. `B_eigen` still is block diagonal.

Comment (recuero) [reply modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C:205]: The test cases for this class `ADHillElastoPlasticityStressUpdate` are not exhaustive and may be overlooking matrix structures that can happen in practice. I'd suggest checking that the assumption is right (hopefully not too compute intensive). If we don't do that, this simplification may lead to wrong results which may be hard to catch in the future.

Comment (SudiptaBiswas) [reply modules/tensor_mechanics/src/materials/ADHillElastoPlasticityStressUpdate.C]: Done!

**Review by recuero — DISMISSED** [review]

**Review by dschwen — DISMISSED** [review]

**Review by dschwen — APPROVED** [review]

**Pull request merged**